text/0097-log-data-model.md

```diff
@@ -0,0 +1,1005 @@
+  
+Both of the above conditions were required to give the field a place in the
+top-level structure of the record.
+
+## Log and Event Record Definition
+
+Note: below we use type `any`, which can be a scalar value (number, string or
+boolean), or an array or map of values. Arbitrary deep nesting of values for
+arrays and maps is allowed (essentially allow to represent an equivalent of a
+JSON object).
+
+[Appendix A](#appendix-a-example-mappings) contains many examples that show how
+existing log formats map to the fields defined below. If there are questions
+about the meaning of the field reviewing the examples may be helpful.
+
+Here is the list of fields in a log record:
+
+Field Name     |Description
+---------------|--------------------------------------------
+Timestamp      |Time when the event occurred.
+TraceId        |Request trace id.
+SpanId         |Request span id.
+TraceFlags     |W3C trace flag.
+SeverityText   |The severity text (also known as log level).
+SeverityNumber |Numerical value of the severity.
+ShortName      |Short event identifier.
+Body           |The body of the log record.
+Resource       |Describes the source of the log.
+Attributes     |Additional information about the event.
+
+Below is the detailed description of each field.
+
+### Field: `Timestamp`
+
+Type: Timestamp, uint64 nanoseconds since Unix epoch.
+
+Description: Time when the event occurred measured by the origin clock. This
+field is optional, it may be missing the timestamp is unknown.
+
+### Trace Context Fields
+
+#### Field: `TraceId`
+
+Type: byte sequence.
+
+Description: Optional request trace id as defined in
+[W3C Trace Context](https://www.w3.org/TR/trace-context/#trace-id). Can be set
+for logs that are part of request processing and have an assigned trace id.
+
+#### Field: `SpanId`
+
+Type: byte sequence.
+
+Description: Optional span id. Can be set for logs that are part of a particular
+processing span. If SpanId is present TraceId SHOULD be also present.
+
+#### Field: `TraceFlags`
+
+Type: byte.
+
+Description: Optional trace flag as defined in
+[W3C Trace Context](https://www.w3.org/TR/trace-context/#trace-flags)
+specification. At the time of writing the specification defines one flag - the
+SAMPLED flag.
+
+### Severity Fields
+
+#### Field: `SeverityText`
+
+Type: string.
+
+Description: the severity text (also known as log level). This is an optional
+field and is the original string representation as it is known at the source. If
+this field is missing and `SeverityNumber` is present then the short name that
+corresponds to the `SeverityNumber` can be used as a substitution.
+
+#### Field: `SeverityNumber`
+
+Type: number.
+
+Description: numerical value of the severity, normalized to values described in
+this document. This is an optional field. If `SeverityNumber` is missing and
+SeverityText is present then it may be assumed that `SeverityNumber` is equal
+to INFO (numeric 9) (see the meaning below).
```

Comment:
Are you suggesting that receiving ends must lookup the value of `SeverityText` in the list of recommended short names and if a match is found assign corresponding `SeverityNumber`? I do not think it is a good approach since it may work in a limited subset of cases and fail in others leading to surprising inconsistency.

For example what should we do if `SeverityText="warning"`? This is not a pre-defined short name for any particular `SeverityNumber`. How would we know it corresponds to WARN(13)?

Even subtler is what we do with case sensitivity, spaces around text, etc, e.g. is "Warn" assumed to be the same as "WARN"?

I think we should play safe and assume that if the source did populate `SeverityText` but did NOT populate `SeverityNumber` then the source did not know how to map severity texts unambiguously. I think it is a safe approach for receivers to treat everything at INFO level. The though process is that everything is "information" unless there is a precise severity level specified for it.

Alternatively, if we don't like this approach then we can just delete this sentence. If `SeverityNumber` is missing then we should not assume any values for it. Whoever reads the record must be ready for missing `SeverityNumber`. This is probably the safest choice that may be will not always result in intuitively best results (e.g. what do we d when comparing `SeverityNumber` to a particular number).

I am inclining towards the alternate approach.

Reply:
I put some thought into this and deleted the sentence which said "it may be assumed that `SeverityNumber` is equal to INFO". This is not data model's business. Data model conveys that the `SeverityNumber` is missing and that's all it needs to convey. Readers of the data can decide what to do in that situation.
Thanks for bringing this up.